File: Week 2/Problem Set/1-readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <stdio.h>
#include <string.h>

// Function prototypes
int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    // Get input text from the user
    string text = get_string("Text: ");

    // Count letters, words, and sentences
    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);

    // Calculate the Coleman-Liau index
    float L = (float) letters / words * 100;
    float S = (float) sentences / words * 100;
    int index = (int) (0.0588 * L - 0.296 * S - 15.8 + 0.5);

    // Output the grade level
    if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %d\n", index);
    }
}

// Count the number of letters in the text
int count_letters(string text)
{
    int count = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isalpha(text[i]))
        {
            count++;
        }
    }
    return count;
}

// Count the number of words in the text
int count_words(string text)
{
    int count = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isspace(text[i]) || i == n - 1)
        {
            count++;
        }
    }
    return count;
}

// Count the number of sentences in the text
int count_sentences(string text)
{
    int count = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == "." || text[i] == "!" || text[i] == "?")
        {
            count++;
        }
    }
    return count;
}

File: Week 6/6-sentimental-readability/readability.py
import re
from cs50 import get_string

while True:
    text = get_string("Text: ")
    if len(text) > 0:
        break


def count_letters():
    return sum(c.isalpha() for c in text)


def count_words():
    return len(re.findall(r'\b[\w\'-]+\b', text))


def count_sentences():
    # If the next character after ".", "!" and "?" is an uppercase letter then it will be counted as a new sentence
    return len(re.split(r'(?<=[.!?])\s+(?=[A-Z])', text))


def main():
    letters = count_letters()
    words = count_words()
    sentences = count_sentences()

    # Calculate the Coleman-Liau index
    L = float(letters / words * 100)
    S = float(sentences / words * 100)
    index = round(0.0588 * L - 0.296 * S - 15.8)

    # Output the grade level
    if (index > 16):
        return ("Grade 16+")
    elif (index < 1):
        return ("Before Grade 1")
    return (f"Grade {index}")


main()
